src/control/control.h

```diff
@@ -36,6 +36,23 @@ class ControlDoublePrivate : public QObject {
     // TODO: don't expose this implementation detail
     constexpr static double kDefaultValue = 0.0;
 
+    struct Parameters {
+        ConfigKey key;
+        ControlObject* pCreatorCO = nullptr;
+        QString name{};
+        QString description{};
+        ControlNumericBehavior* behavior = nullptr;
+        double defaultValue = kDefaultValue;
+        bool ignoreNops = true;
+        bool track = false;
+        bool persist = false;
+        bool keyboardRepeatable = false;
+    };
+
+    struct ParametersWithFlags : public Parameters {
```

Comment:
I have not yet fully understand why we have two inherited types. Maybe we can consider to merge them in a next step? Maybe we can even make some more booleans flags.

Reply:
because the downstream APIs want to take more parameters. Taking those parameters as separate function parameters would defeat the purpose again (taking a single struct so they are ordering independent and can act as named arguments). Merging all the downstream options into the upstream (parent/base) struct would make the API confusing because the caller can now pass options which don't make sense in the context of the upstream API. 
Merging these downstream options with the upstream parameters would violate "separation of concerns" and the "open-closed principle". 

Does it help if you interpret the public inheritance simply as a set union operation ( eg, `ParametersWithFlags` $=$`Parameters` $\cup$ `Flags`)?

Reply:
Rereading this and addressing your original question:
we have `ParametersWithConfirm` because that what the private `ControlDoublePrivate` ctor needs and we have `ParametersWithFlags` because thats what `ControlDoublePrivate::getControl` wants. These are separate because `ParametersWithConfirm` doesn't care about the extra flags in `ParametersWithFlags` while the public `ParametersWithFlags` should not expose the internal `confirmRequired` of `ParametersWithConfirm`. Merging these two doesn't make sense because they're supposed to be used in different contexts.